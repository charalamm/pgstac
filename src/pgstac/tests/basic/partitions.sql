SET pgstac.use_queue=TRUE;
--test non-partitioned collection
INSERT INTO collections (content) VALUES ('{"id":"pgstactest-partitioned"}');
INSERT INTO items_staging_ignore(content)
SELECT jsonb_build_object(
    'id', concat('pgstactest-partitioned-', (row_number() over ())::text),
    'collection', 'pgstactest-partitioned',
    'geometry', '{"type": "Polygon", "coordinates": [[[-85.309412, 30.933949], [-85.308201, 31.002658], [-85.378084, 31.003555], [-85.379245, 30.934843], [-85.309412, 30.933949]]]}'::json,
    'properties', jsonb_build_object( 'datetime', g::text)
) FROM generate_series('2020-01-01'::timestamptz, '2022-01-01'::timestamptz, '1 day'::interval) g;
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned';

--test collection partioned by year
INSERT INTO collections (content, partition_trunc) VALUES ('{"id":"pgstactest-partitioned-year"}', 'year');
INSERT INTO items_staging_ignore(content)
SELECT jsonb_build_object(
    'id', concat('pgstactest-partitioned-year', (row_number() over ())::text),
    'collection', 'pgstactest-partitioned-year',
    'geometry', '{"type": "Polygon", "coordinates": [[[-85.309412, 30.933949], [-85.308201, 31.002658], [-85.378084, 31.003555], [-85.379245, 30.934843], [-85.309412, 30.933949]]]}'::json,
    'properties', jsonb_build_object( 'datetime', g::text)
) FROM generate_series('2020-01-01'::timestamptz, '2022-01-01'::timestamptz, '1 day'::interval) g;
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-year';

--test collection partioned by month
INSERT INTO collections (content, partition_trunc) VALUES ('{"id":"pgstactest-partitioned-month"}', 'month');
INSERT INTO items_staging_ignore(content)
SELECT jsonb_build_object(
    'id', concat('pgstactest-partitioned-month', (row_number() over ())::text),
    'collection', 'pgstactest-partitioned-month',
    'geometry', '{"type": "Polygon", "coordinates": [[[-85.309412, 30.933949], [-85.308201, 31.002658], [-85.378084, 31.003555], [-85.379245, 30.934843], [-85.309412, 30.933949]]]}'::json,
    'properties', jsonb_build_object( 'datetime', g::text)
) FROM generate_series('2020-01-01'::timestamptz, '2022-01-01'::timestamptz, '1 day'::interval) g;
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-month';

--test repartitioning from year to non partitioned
UPDATE collections SET partition_trunc=NULL WHERE id='pgstactest-partitioned-year';
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-year';
SELECT count(*) FROM items WHERE collection='pgstactest-partitioned-year';

--test repartitioning from non-partitioned to year
UPDATE collections SET partition_trunc='year' WHERE id='pgstactest-partitioned';
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned';
SELECT count(*) FROM items WHERE collection='pgstactest-partitioned';

--test noop for repartitioning
UPDATE collections SET content=content || '{"foo":"bar"}'::jsonb WHERE id='pgstactest-partitioned-month';
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-month';
SELECT count(*) FROM items WHERE collection='pgstactest-partitioned-month';

--check that partition stats haven't been updated
SELECT collection, partition_dtrange, constraint_dtrange, constraint_edtrange, dtrange, edtrange, spatial FROM partitions WHERE collection='pgstactest-partitioned-year';

--check that queue has items
SELECT count(*) FROM query_queue;

--run queue items to update partition stats
SELECT run_queued_queries_intransaction();

--check that partition stats have been updated
SELECT collection, partition_dtrange, constraint_dtrange, constraint_edtrange, dtrange, edtrange, spatial FROM partitions WHERE collection='pgstactest-partitioned-year';
