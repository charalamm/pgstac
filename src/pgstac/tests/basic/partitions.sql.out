SET pgstac.use_queue=TRUE;
SET
--test non-partitioned collection
INSERT INTO collections (content) VALUES ('{"id":"pgstactest-partitioned"}');
INSERT 0 1
INSERT INTO items_staging_ignore(content)
SELECT jsonb_build_object(
    'id', concat('pgstactest-partitioned-', (row_number() over ())::text),
    'collection', 'pgstactest-partitioned',
    'geometry', '{"type": "Polygon", "coordinates": [[[-85.309412, 30.933949], [-85.308201, 31.002658], [-85.378084, 31.003555], [-85.379245, 30.934843], [-85.309412, 30.933949]]]}'::json,
    'properties', jsonb_build_object( 'datetime', g::text)
) FROM generate_series('2020-01-01'::timestamptz, '2022-01-01'::timestamptz, '1 day'::interval) g;
INSERT 0 732
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned';
     1

--test collection partioned by year
INSERT INTO collections (content, partition_trunc) VALUES ('{"id":"pgstactest-partitioned-year"}', 'year');
INSERT 0 1
INSERT INTO items_staging_ignore(content)
SELECT jsonb_build_object(
    'id', concat('pgstactest-partitioned-year', (row_number() over ())::text),
    'collection', 'pgstactest-partitioned-year',
    'geometry', '{"type": "Polygon", "coordinates": [[[-85.309412, 30.933949], [-85.308201, 31.002658], [-85.378084, 31.003555], [-85.379245, 30.934843], [-85.309412, 30.933949]]]}'::json,
    'properties', jsonb_build_object( 'datetime', g::text)
) FROM generate_series('2020-01-01'::timestamptz, '2022-01-01'::timestamptz, '1 day'::interval) g;
INSERT 0 732
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-year';
     3

--test collection partioned by month
INSERT INTO collections (content, partition_trunc) VALUES ('{"id":"pgstactest-partitioned-month"}', 'month');
INSERT 0 1
INSERT INTO items_staging_ignore(content)
SELECT jsonb_build_object(
    'id', concat('pgstactest-partitioned-month', (row_number() over ())::text),
    'collection', 'pgstactest-partitioned-month',
    'geometry', '{"type": "Polygon", "coordinates": [[[-85.309412, 30.933949], [-85.308201, 31.002658], [-85.378084, 31.003555], [-85.379245, 30.934843], [-85.309412, 30.933949]]]}'::json,
    'properties', jsonb_build_object( 'datetime', g::text)
) FROM generate_series('2020-01-01'::timestamptz, '2022-01-01'::timestamptz, '1 day'::interval) g;
INSERT 0 732
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-month';
    25

--test repartitioning from year to non partitioned
UPDATE collections SET partition_trunc=NULL WHERE id='pgstactest-partitioned-year';
UPDATE 1
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-year';
     1

SELECT count(*) FROM items WHERE collection='pgstactest-partitioned-year';
   732

--test repartitioning from non-partitioned to year
UPDATE collections SET partition_trunc='year' WHERE id='pgstactest-partitioned';
UPDATE 1
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned';
     3

SELECT count(*) FROM items WHERE collection='pgstactest-partitioned';
   732

--test noop for repartitioning
UPDATE collections SET content=content || '{"foo":"bar"}'::jsonb WHERE id='pgstactest-partitioned-month';
UPDATE 1
SELECT count(*) FROM partitions WHERE collection='pgstactest-partitioned-month';
    25

SELECT count(*) FROM items WHERE collection='pgstactest-partitioned-month';
   732

--check that partition stats haven't been updated
SELECT collection, partition_dtrange, constraint_dtrange, constraint_edtrange, dtrange, edtrange, spatial FROM partitions WHERE collection='pgstactest-partitioned-year';
pgstactest-partitioned-year | [-infinity,infinity] | [-infinity,infinity] | [-infinity,infinity] |         |          |

--check that queue has items
SELECT count(*) FROM query_queue;
    99

--run queue items to update partition stats
SELECT run_queued_queries_intransaction();
                               99

--check that partition stats have been updated
SELECT collection, partition_dtrange, constraint_dtrange, constraint_edtrange, dtrange, edtrange, spatial FROM partitions WHERE collection='pgstactest-partitioned-year';
 pgstactest-partitioned-year | [-infinity,infinity] | ["2020-01-01 00:00:00+00","2022-01-01 00:00:00+00"] | ["2020-01-01 00:00:00+00","2022-01-01 00:00:00+00"] | ["2020-01-01 00:00:00+00","2022-01-01 00:00:00+00"] | ["2020-01-01 00:00:00+00","2022-01-01 00:00:00+00"] | 01030000000100000005000000000000A0455855C00000004017EF3E40000000A0455855C000000000E9003F4000000080B95355C000000000E9003F4000000080B95355C00000004017EF3E40000000A0455855C00000004017EF3E40

\set QUIET 1
\set ECHO none
